File: app/src/main/java/org/gnucash/android/ui/colorpicker/ColorPickerPalette.java
/*
 * Copyright (C) 2013 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.gnucash.android.ui.colorpicker;


import static java.lang.Math.max;

import android.content.Context;
import android.content.res.Resources;
import android.util.AttributeSet;
import android.view.View;
import android.widget.GridLayout;

import androidx.annotation.ColorInt;

import org.gnucash.android.R;
import org.gnucash.android.ui.colorpicker.ColorPickerSwatch.OnColorSelectedListener;

import java.util.ArrayList;
import java.util.List;

/**
 * A color picker custom view which creates an grid of color squares.  The number of squares per
 * row (and the padding between the squares) is determined by the user.
 */
public class ColorPickerPalette extends GridLayout {

    public static final int SIZE_LARGE = 1;
    public static final int SIZE_SMALL = 2;

    private OnColorSelectedListener mOnColorSelectedListener;

    private String mDescription;
    private String mDescriptionSelected;

    private int mSwatchLength;
    private int mMarginSize;
    private int mNumColumns = UNDEFINED;
    private final List<ColorPickerSwatch> swatches = new ArrayList<>();

    public ColorPickerPalette(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public ColorPickerPalette(Context context) {
        super(context);
    }

    /**
     * Initialize the size, columns, and listener.  Size should be a pre-defined size (SIZE_LARGE
     * or SIZE_SMALL) from ColorPickerDialogFragment.
     */
    public void init(int size, int columns, OnColorSelectedListener listener) {
        if (columns > 0) {
            mNumColumns = columns;
            setColumnCount(columns);
        }

        Resources res = getResources();
        if (size == SIZE_LARGE) {
            mSwatchLength = res.getDimensionPixelSize(R.dimen.color_swatch_large);
            mMarginSize = res.getDimensionPixelSize(R.dimen.color_swatch_margins_large);
        } else {
            mSwatchLength = res.getDimensionPixelSize(R.dimen.color_swatch_small);
            mMarginSize = res.getDimensionPixelSize(R.dimen.color_swatch_margins_small);
        }
        mOnColorSelectedListener = listener;

        mDescription = res.getString(R.string.color_swatch_description);
        mDescriptionSelected = res.getString(R.string.color_swatch_description_selected);
    }

    /**
     * Adds swatches to table in a serpentine format.
     */
    public void drawPalette(int[] colors, int selectedColor) {
        if (colors == null) {
            return;
        }

        swatches.clear();
        removeAllViews();

        int tableElements = 0;

        // Fills the table with swatches based on the array of colors.
        for (int color : colors) {
            tableElements++;

            ColorPickerSwatch colorSwatch = createColorSwatch(color, selectedColor);
            swatches.add(colorSwatch);
            setSwatchDescription(tableElements, color == selectedColor, colorSwatch);
            addView(colorSwatch);
        }
    }

    /**
     * Add a content description to the specified swatch view. Because the colors get added in a
     * snaking form, every other row will need to compensate for the fact that the colors are added
     * in an opposite direction from their left->right/top->bottom order, which is how the system
     * will arrange them for accessibility purposes.
     */
    private void setSwatchDescription(int index, boolean selected, View swatch) {
        final String description;
        if (selected) {
            description = String.format(mDescriptionSelected, index);
        } else {
            description = String.format(mDescription, index);
        }
        swatch.setContentDescription(description);
    }

    /**
     * Creates a color swatch.
     */
    private ColorPickerSwatch createColorSwatch(@ColorInt int color, @ColorInt int selectedColor) {
        ColorPickerSwatch view = new ColorPickerSwatch(getContext(), color,
            color == selectedColor, mOnColorSelectedListener);
        view.setLayoutParams(generateSwatchParams());
        return view;
    }

    private LayoutParams generateSwatchParams() {
        LayoutParams params = generateDefaultLayoutParams();
        params.width = mSwatchLength;
        params.height = mSwatchLength;
        params.setMargins(mMarginSize, mMarginSize, mMarginSize, mMarginSize);
        return params;
    }

    public void setSelected(@ColorInt int color) {
        for (ColorPickerSwatch swatch : swatches) {
            swatch.setChecked(swatch.color == color);
        }
    }

    @Override
    protected void onMeasure(int widthSpec, int heightSpec) {
        super.onMeasure(widthSpec, heightSpec);

        if (mNumColumns < 0) {
            int hPadding = getPaddingLeft() + getPaddingRight();
            final int widthSansPadding = getMeasuredWidth() - hPadding;
            int widthSwatch = mMarginSize + mSwatchLength + mMarginSize;
            int columnCount = max(1, widthSansPadding / widthSwatch);
            int count = getChildCount();
            // Reset the layout params to recalculate their spans.
            for (int i = 0; i < count; i++) {
                View child = getChildAt(i);
                child.setLayoutParams(generateSwatchParams());
            }
            setColumnCount(columnCount);
        }
    }
}

File: app/src/main/java/org/gnucash/android/ui/colorpicker/ColorPickerPalette.kt
/*
 * Copyright (C) 2013 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.gnucash.android.ui.colorpicker

import android.content.Context
import android.content.res.Configuration
import android.graphics.Color
import android.util.AttributeSet
import android.view.ViewGroup
import androidx.annotation.ColorInt
import androidx.core.view.setPadding
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import org.gnucash.android.R
import org.gnucash.android.ui.colorpicker.ColorPickerSwatch.OnColorSelectedListener
import kotlin.math.max
import kotlin.math.min

/**
 * A color picker custom view which creates an grid of color squares.  The number of squares per
 * row (and the padding between the squares) is determined by the user.
 */
class ColorPickerPalette @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null) :
    RecyclerView(context, attrs) {
    private var onColorSelectedListener: OnColorSelectedListener? = null
    private var swatchSizePx = 0
    private var marginSizePx = 0
    private var spanCount = UNDEFINED
    private var swatchAdapter: ColorPickerAdapter? = null

    /**
     * Initialize the size, columns, and listener.
     *
     * @param size  Size should be a pre-defined size (`ColorPickerPalette.SIZE_LARGE` or `ColorPickerPalette.SIZE_SMALL`)
     * @param columnCount the maximum number of columns.
     * @param listener the listener for when the color is selected.
     */
    fun init(size: Int, spanCount: Int, listener: OnColorSelectedListener?) {
        this.spanCount = if (spanCount >= 1) spanCount else UNDEFINED

        val res = resources
        applyOrientation(res.configuration.orientation)

        if (size == SIZE_LARGE) {
            swatchSizePx = res.getDimensionPixelSize(R.dimen.color_swatch_large)
            marginSizePx = res.getDimensionPixelSize(R.dimen.color_swatch_margins_large)
        } else {
            swatchSizePx = res.getDimensionPixelSize(R.dimen.color_swatch_small)
            marginSizePx = res.getDimensionPixelSize(R.dimen.color_swatch_margins_small)
        }
        onColorSelectedListener = listener
    }

    /**
     * Adds swatches to table in a serpentine format.
     */
    fun drawPalette(colors: IntArray?, colorNames: Array<String>?, @ColorInt selectedColor: Int) {
        if (colors == null) {
            return
        }

        swatchAdapter =
            ColorPickerAdapter(colors, colorNames?.toList() ?: emptyList(), selectedColor)
        adapter = swatchAdapter
    }

    fun setSelected(@ColorInt color: Int) {
        val adapter = swatchAdapter ?: return
        for (swatch in adapter.swatches) {
            swatch.isChecked = swatch.color == color
        }
    }

    override fun onMeasure(widthSpec: Int, heightSpec: Int) {
        super.onMeasure(widthSpec, heightSpec)

        if (spanCount < 1) {
            val count: Int
            val layoutManager = layoutManager as GridLayoutManager
            if (layoutManager.orientation == GridLayoutManager.VERTICAL) {
                val hPadding = paddingLeft + paddingRight
                val widthSansPadding = measuredWidth - hPadding
                val widthSwatch = marginSizePx + swatchSizePx + marginSizePx
                count = max(1, (widthSansPadding / widthSwatch))
            } else {
                val vPadding = paddingTop + paddingBottom
                val heightSansPadding = measuredHeight - vPadding
                val heightSwatch = marginSizePx + swatchSizePx + marginSizePx
                count = max(1, (heightSansPadding / heightSwatch))
            }
            layoutManager.spanCount = count
        }
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        applyOrientation(newConfig.orientation)
    }

    private fun applyOrientation(orientation: Int) {
        val layoutManager = object : GridLayoutManager(context, max(1, spanCount)) {
            override fun checkLayoutParams(lp: RecyclerView.LayoutParams): Boolean {
                if (!super.checkLayoutParams(lp)) return false

                val sizeGrid = if (orientation == VERTICAL) {
                    measuredWidth - paddingStart - paddingEnd
                } else {
                    measuredHeight - paddingTop - paddingBottom
                }
                var sizeCell = min(swatchSizePx, max(0, sizeGrid / spanCount))
                if (sizeCell == 0) {
                    sizeCell = swatchSizePx
                }
                lp.width = sizeCell
                lp.height = sizeCell
                return true
            }
        }
        if (orientation == Configuration.ORIENTATION_LANDSCAPE) {
            layoutManager.orientation = GridLayoutManager.HORIZONTAL
        } else {
            layoutManager.orientation = GridLayoutManager.VERTICAL
        }
        this.layoutManager = layoutManager
    }

    private class ColorPickerViewHolder(private val swatch: ColorPickerSwatch) :
        ViewHolder(swatch) {
        fun bind(
            position: Int,
            @ColorInt color: Int,
            colorName: String?,
            @ColorInt selectedColor: Int
        ) {
            val selected = color == selectedColor
            swatch.setColor(color)
            swatch.isChecked = selected

            val context: Context = itemView.context
            val name = colorName ?: (position + 1).toString()
            swatch.contentDescription = if (selected) {
                context.getString(R.string.color_swatch_description_selected, name)
            } else {
                context.getString(R.string.color_swatch_description, name)
            }
        }
    }

    private inner class ColorPickerAdapter(
        private val colors: IntArray,
        private val colorNames: List<String>,
        @field:ColorInt
        private val selectedColor: Int = Color.TRANSPARENT
    ) : Adapter<ColorPickerViewHolder>() {
        private val _swatches = mutableListOf<ColorPickerSwatch>()
        val swatches: List<ColorPickerSwatch> = _swatches

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ColorPickerViewHolder {
            val swatch = createColorSwatch(parent)
            _swatches.add(swatch)
            return ColorPickerViewHolder(swatch)
        }

        override fun onBindViewHolder(holder: ColorPickerViewHolder, position: Int) {
            @ColorInt val color = colors[position]
            val colorName: String? = if (position < colorNames.size) {
                colorNames[position]
            } else {
                null
            }
            holder.bind(position, color, colorName, selectedColor)
        }

        override fun getItemCount(): Int {
            return colors.size
        }

        /**
         * Creates a color swatch.
         */
        private fun createColorSwatch(parent: ViewGroup): ColorPickerSwatch {
            val view = ColorPickerSwatch(
                context = parent.context,
                onColorSelectedListener = onColorSelectedListener
            )
            view.setPadding(marginSizePx)
            return view
        }
    }

    companion object {
        const val SIZE_LARGE: Int = 1
        const val SIZE_SMALL: Int = 2

        private const val UNDEFINED = Int.MIN_VALUE
    }
}